File: 240703/돌 잘 치우기/clear-stones-well.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>
using namespace std;

#define MAX_N 100
#define DIR_NUM 4

int n, k, m;
int grid[MAX_N][MAX_N];

int ans;

vector<pair<int, int>> s_pos;
vector<pair<int, int>> stone_pos;
vector<pair<int, int>> selected_stones;

queue<pair<int, int>> q;
bool visited[MAX_N][MAX_N];

bool InRange(int x, int y) {
    return 0<=x && x<n && 0<=y && y<n;
}

bool CanGo(int x, int y) {
    if(!InRange(x, y)) return false;
    if(visited[x][y] || grid[x][y]) return false;

    return true;
}

void BFS() {
    while(!q.empty()) {
        int x, y;
        tie(x, y) = q.front();
        q.pop();

        int dx[DIR_NUM] = {1, -1, 0, 0};
        int dy[DIR_NUM] = {0, 0, 1, -1};

        for(int d=0; d<DIR_NUM; d++) {
            int nx = x+dx[d], ny=y+dy[d];

            if(CanGo(nx, ny)) {
                q.push(make_pair(nx, ny));
                visited[nx][ny] = true;
            }
        }
    }
}

int Calc() {
    for(int i=0; i<m; i++) {
        int x, y;
        tie(x, y) = selected_stones[i];
        grid[x][y] = 0;
    }

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            visited[i][j] = 0;
        }
    }

    for(int i=0; i<k; i++) {
        q.push(s_pos[i]);
        visited[s_pos[i].first][s_pos[i].second] = true;
    }

    BFS();

    for(int i=0; i<m; i++) {
        int x, y;
        tie(x, y) = selected_stones[i];
        grid[x][y] = 1;
    }

    int cnt = 0;
    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            if(visited[i][j])
                cnt++;
        }
    }

    return cnt;
}

void FindMax(int idx, int cnt) {
    if(idx == (int) stone_pos.size()) {
        if(cnt == m)
            ans = max(ans, Calc());
        return;
    }

    selected_stones.push_back(stone_pos[idx]);
    FindMax(idx+1, cnt+1);
    selected_stones.pop_back();

    FindMax(idx+1, cnt);
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n>>k>>m;

    for(int i=0; i<n; i++) {
        for(int j=0; j<n; j++) {
            cin >> grid[i][j];
            if(grid[i][j])
                stone_pos.push_back(make_pair(i, j));
        }
    }

    for(int i=0; i<k; i++) {
        int r, c;
        cin >> r>>c;
        s_pos.push_back(make_pair(r-1, c-1));
    }

    FindMax(0, 0);
    cout << ans;
    return 0;
}

File: 241101/돌 잘 치우기/clear-stones-well.py
from collections import deque

n, k, m = tuple(map(int, input().split()))
grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

ans = 0
s_pos = []
stone_pos = [
    (i, j)
    for i in range(n)
    for j in range(n)
    if grid[i][j] == 1
]

selected_stones = []

q = deque()
visited = [
    [0 for _ in range(n)]
    for _ in range(n)
]

def canGo(x, y):
    return 0<=x<n and 0<=y<n and grid[x][y] == 0 and not visited[x][y]

def bfs():
    dxs, dys = [-1, 0, 1, 0], [0, 1, 0, -1]

    while q:
        x, y = q.popleft()

        for dx, dy in zip(dxs, dys):
            nx, ny = x+dx, y+dy

            if canGo(nx, ny):
                q.append((nx, ny))
                visited[nx][ny] = 1

def Calc():
    for x, y in selected_stones:
        grid[x][y] = 0
    
    for i in range(n):
        for j in range(n):
            visited[i][j] = 0

    for x, y in s_pos:
        q.append((x, y))
        visited[x][y] = 1
    
    bfs()

    for x, y in selected_stones:
        grid[x][y] = 1

    cnt = 0
    for i in range(n):
        for j in range(n):
            if visited[i][j]:
                cnt += 1
    
    return cnt

def find_max(idx, cnt):
    global ans

    if idx == len(stone_pos):
        if cnt == m:
            ans = max(ans, Calc())
        return
    
    selected_stones.append(stone_pos[idx])
    find_max(idx+1, cnt+1)
    selected_stones.pop()

    find_max(idx+1, cnt)

for _ in range(k):
    r, c = tuple(map(int, input().split()))
    s_pos.append((r-1, c-1))

find_max(0, 0)
print(ans)